libyul/optimiser/StackLimitEvader.cpp

```diff
@@ -0,0 +1,156 @@
+//   - Visit all children that are not already visited.
+//   - Determe the maximum value ``n`` of the values of ``nextAvailableSlot`` among the children.
+//   - If the function itself contains variables that need memory slots, but is contained in a cycle,
+//     abort the process as failure.
+//   - If not, assign each variable its slot starting starting from ``n`` (incrementing it).
+//   - Assign ``nextAvailableSlot`` of the function to ``n``.
+struct MemoryOffsetAllocator
+{
+	map<YulString, set<YulString>> const& unreachableVariables;
+	map<YulString, set<YulString>> const& callGraph;
+
+	uint64_t run(YulString _function = YulString{})
+	{
+		if (nextAvailableSlot.count(_function))
+			return nextAvailableSlot[_function];
+
+		// Assign to zero early to guard against recursive calls.
+		nextAvailableSlot[_function] = 0;
+
+		uint64_t nextSlot = 0;
+		if (callGraph.count(_function))
+			for (auto child: callGraph.at(_function))
+				nextSlot = std::max(run(child), nextSlot);
+
+		if (unreachableVariables.count(_function))
+		{
+			yulAssert(!slotAllocations.count(_function), "");
+			auto& assignedSlots = slotAllocations[_function];
+			for (auto const& variable: unreachableVariables.at(_function))
+				if (variable.empty())
+				{
+					// TODO: Too many function arguments or return parameters.
+				}
+				else
+					assignedSlots[variable] = nextSlot++;
+		}
+
+		return (nextAvailableSlot[_function] = nextSlot);
+	}
+
+	map<YulString, map<YulString, uint64_t>> slotAllocations{};
+	map<YulString, uint64_t> nextAvailableSlot{};
+};
+}
+
+void StackLimitEvader::run(OptimiserStepContext& _context, Object& _object, bool _optimizeStackAllocation)
+{
+	run(_context, _object, CompilabilityChecker(
+		_context.dialect,
+		_object,
+		_optimizeStackAllocation
+	).unreachableVariables);
+}
+
+void StackLimitEvader::run(
+	OptimiserStepContext& _context,
+	Object& _object,
+	std::map<YulString, std::set<YulString>> const& _unreachableVariables)
+{
+	yulAssert(_object.code, "");
+	auto const* evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+	yulAssert(
+		evmDialect && evmDialect->providesObjectAccess(),
+		"StackToMemoryMover can only be run on objects using the EVMDialect with object access."
+	);
+
+	// Find the literal argument of the ``memoryinit`` call, if there is a unique such call, otherwise abort.
+	Literal* memoryInitLiteral = nullptr;
+	if (
+		auto memoryInits = FunctionCallFinder::run(*_object.code, "memoryinit"_yulstring);
+		memoryInits.size() == 1
+	)
+		memoryInitLiteral = std::get_if<Literal>(&memoryInits.front()->arguments.back());
+	if (!memoryInitLiteral)
+		return;
+	u256 reservedMemory = valueOfLiteral(*memoryInitLiteral);
+
+	map<YulString, set<YulString>> callGraph = CallGraphGenerator::callGraph(*_object.code).functionCalls;
+
+	// Collect all names of functions contained in cycles in the callgraph.
+	// TODO: this algorithm is suboptimal and can be improved. It also overlaps with Semantics.cpp.
```

Comment:
Actually this *doesn't* exactly overlap with Semantics.cpp.
Semantics.cpp needs to know if a function *contains* a cycle. Here I need to know if a function *is contained in* a cycle.

I.e.
```
function f() {
  g()
}
function g() { h() }
function h() { g() }
```
The algorithm in ``Semantics.cpp`` considers ``f`` as recursive, however here I don't want that.

However, the other way round works, I can have ``Semantics.cpp`` use "contained in a cycle" instead, since it will propagate side effects up the callgraph anyways... so I think I'll extract this one in a slightly nicer form and have Semantics.cpp use it.